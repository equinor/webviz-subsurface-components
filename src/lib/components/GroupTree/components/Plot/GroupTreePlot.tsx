import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React, { useMemo } from "react";
import { useResizeDetector } from "react-resize-detector";
import { FlowRate, Node } from "../../redux/types";
import "./dynamic_tree.css";
import { getLayout, Padding } from "./plotUtil";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            display: "flex",
            flex: 1,
            height: "80%",
            padding: theme.spacing(1),
        },
    })
);
interface Props {
    root: Node;
    currentFlowRate: FlowRate;
}

const padding: Padding = { left: 80, right: 50, top: 70, bottom: 50 };
/* eslint-disable react/prop-types */
const GroupTreePlot: React.FC<Props> = React.memo(
    ({ root, currentFlowRate }) => {
        const classes = useStyles();
        const { width, height, ref } = useResizeDetector({
            refreshMode: "debounce",
            refreshRate: 50,
            refreshOptions: { trailing: true },
        });
        const layout = useMemo(
            () =>
                width !== undefined && height !== undefined
                    ? getLayout(width, height, padding)
                    : undefined,
            [width, height]
        );

        return (
            <div
                className={classes.root}
                ref={ref as React.LegacyRef<HTMLDivElement>}
                data-tip
                data-for="plot-tooltip"
            >
                {layout && (
                    <svg id={"svg-context"} width={width} height={height}>
                        <g
                            transform={`translate(${padding.left},${padding.top})`}
                        ></g>
                    </svg>
                )}
            </div>
        );
    }
);

export default GroupTreePlot;
