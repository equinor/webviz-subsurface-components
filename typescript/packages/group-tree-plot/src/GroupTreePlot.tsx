import React from "react";

import GroupTreeAssembler from "./GroupTreeAssembler/groupTreeAssembler";
import type { DatedTree, EdgeMetadata, NodeMetadata } from "./types";
import { isEqual } from "lodash";

export interface GroupTreePlotProps {
    id: string;
    edgeMetadataList: EdgeMetadata[];
    nodeMetadataList: NodeMetadata[];
    datedTrees: DatedTree[];
    selectedEdgeKey: string;
    selectedNodeKey: string;
    selectedDateTime: string;
}

export const GroupTreePlot: React.FC<GroupTreePlotProps> = (
    props: GroupTreePlotProps
) => {
    const divRef = React.useRef<HTMLDivElement>(null);
    const groupTreeAssemblerRef = React.useRef<GroupTreeAssembler>();

    // State to ensure divRef is defined before creating GroupTree
    const [isMounted, setIsMounted] = React.useState<boolean>(false);

    // Remove when typescript version is implemented using ref
    const [prevId, setPrevId] = React.useState<string | null>(null);

    const [prevDatedTrees, setPrevDatedTrees] = React.useState<
        DatedTree[] | null
    >(null);

    const [prevSelectedEdgeKey, setPrevSelectedEdgeKey] =
        React.useState<string>(props.selectedEdgeKey);
    const [prevSelectedNodeKey, setPrevSelectedNodeKey] =
        React.useState<string>(props.selectedNodeKey);
    const [prevSelectedDateTime, setPrevSelectedDateTime] =
        React.useState<string>(props.selectedDateTime);

    React.useEffect(function initialRender() {
        setIsMounted(true);
    }, []);

    if (
        isMounted &&
        divRef.current &&
        (!isEqual(prevDatedTrees, props.datedTrees) ||
            prevId !== divRef.current.id)
    ) {
        setPrevDatedTrees(props.datedTrees);
        setPrevId(divRef.current.id);
        groupTreeAssemblerRef.current = new GroupTreeAssembler(
            divRef.current.id,
            props.datedTrees,
            props.selectedEdgeKey,
            props.selectedNodeKey,
            props.selectedDateTime,
            props.edgeMetadataList,
            props.nodeMetadataList
        );
    }

    if (prevSelectedEdgeKey !== props.selectedEdgeKey) {
        setPrevSelectedEdgeKey(props.selectedEdgeKey);
        if (groupTreeAssemblerRef.current) {
            groupTreeAssemblerRef.current.flowrate = props.selectedEdgeKey;
        }
    }

    if (prevSelectedNodeKey !== props.selectedNodeKey) {
        setPrevSelectedNodeKey(props.selectedNodeKey);
        if (groupTreeAssemblerRef.current) {
            groupTreeAssemblerRef.current.nodeinfo = props.selectedNodeKey;
        }
    }

    if (prevSelectedDateTime !== props.selectedDateTime) {
        setPrevSelectedDateTime(props.selectedDateTime);
        if (groupTreeAssemblerRef.current) {
            groupTreeAssemblerRef.current.update(props.selectedDateTime);
        }
    }

    return <div id={props.id} ref={divRef} />;
};

GroupTreePlot.displayName = "GroupTreePlot";
