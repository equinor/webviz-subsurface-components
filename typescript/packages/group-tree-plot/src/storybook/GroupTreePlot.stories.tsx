import type { Meta, StoryObj } from "@storybook/react";
import React from "react";

import { GroupTreePlot } from "../GroupTreePlot";

import type { EdgeMetadata, NodeMetadata } from "../types";

import {
    exampleDatedTrees,
    exampleDates,
} from "../../example-data/dated-trees";

const stories: Meta = {
    component: GroupTreePlot,
    title: "GroupTreePlot/Demo",
    argTypes: {
        selectedDateTime: {
            description:
                "The selected `string` must be a date time present in one of the `dates` arrays in an element of the`datedTrees`-prop.\n\n",
        },
        selectedEdgeKey: {
            description:
                "The selection `string` must be an edge key present in one of the `edge_data` objects in the `tree`-prop of an element in `datedTrees`-prop.\n\n",
        },
        selectedNodeKey: {
            description:
                "The selected `string` must be a node key present in one of the `node_data` objects in the `tree`-prop of an element in `datedTrees`-prop.\n\n",
        },
    },
};
export default stories;

/**
 * Storybook test for the group tree plot component
 */

const Template = (args) => {
    return (
        <GroupTreePlot
            id={args.id}
            datedTrees={args.datedTrees}
            edgeMetadataList={args.edgeMetadataList}
            nodeMetadataList={args.nodeMetadataList}
            selectedDateTime={args.selectedDateTime}
            selectedEdgeKey={args.selectedEdgeKey}
            selectedNodeKey={args.selectedNodeKey}
        />
    );
};

const edgeMetadataList: EdgeMetadata[] = [
    { key: "waterrate", label: "Water Rate", unit: "m3/s" },
    { key: "oilrate", label: "Oil Rate", unit: "m3/s" },
    { key: "gasrate", label: "Gas Rate", unit: "m3/s" },
    { key: "waterinjrate", label: "Water Injection Rate", unit: "m3/s" },
    { key: "gasinjrate", label: "Gas Injection Rate", unit: "m3/s" },
];

const nodeMetadataList: NodeMetadata[] = [
    { key: "pressure", label: "Pressure", unit: "Bar" },
    { key: "bhp", label: "Bottom Hole Pressure", unit: "N/m2" },
    { key: "wmctl", label: "Missing label", unit: "Unknown unit" },
];

export const Default: StoryObj<typeof Template> = {
    args: {
        id: "grouptreeplot",
        datedTrees: exampleDatedTrees,
        edgeMetadataList: edgeMetadataList,
        nodeMetadataList: nodeMetadataList,
        selectedDateTime: exampleDates[0],
        selectedEdgeKey: edgeMetadataList[0].key,
        selectedNodeKey: nodeMetadataList[0].key,
    },
    render: (args) => <Template {...args} />,
};
