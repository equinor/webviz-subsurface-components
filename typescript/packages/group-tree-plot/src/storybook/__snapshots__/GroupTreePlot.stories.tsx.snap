// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupTreePlot/Demo Default smoke-test 2`] = `
<div style="width: 938px; height: 700px;">
  <svg height="100%"
       width="100%"
  >
    <g transform="translate(70,10)">
      <g>
        <path id=":r1:"
              class="link grouptree_link grouptree_link__waterrate"
              stroke-dasharray="none"
              stroke-opacity="1"
              stroke-width="57.99999999999999"
              d="
        M 0 340         C             374 340,             374 170,             748 170
        "
        >
          <title>
            Water Rate: 20 m3/s
Oil Rate: 30 m3/s
Gas Rate: 40 m3/s
Water Injection Rate: 50 m3/s
Gas Injection Rate: 60 m3/s
          </title>
        </path>
        <text dominant-baseline="central"
              text-anchor="middle"
              fill-opacity="1"
        >
          <textPath class="edge_info_text"
                    startoffset="50%"
                    href="#:r1:"
          >
            VFP11
          </textPath>
        </text>
      </g>
      <g>
        <path id=":r3:"
              class="link grouptree_link grouptree_link__waterrate"
              stroke-dasharray="none"
              stroke-opacity="1"
              stroke-width="72"
              d="
        M 0 340         C             374 340,             374 510,             748 510
        "
        >
          <title>
            Water Rate: 25 m3/s
Oil Rate: 35 m3/s
Gas Rate: 45 m3/s
Water Injection Rate: 55 m3/s
Gas Injection Rate: 65 m3/s
          </title>
        </path>
        <text dominant-baseline="central"
              text-anchor="middle"
              fill-opacity="1"
        >
          <textPath class="edge_info_text"
                    startoffset="50%"
                    href="#:r3:"
          >
            VFP12
          </textPath>
        </text>
      </g>
      <g class="node"
         cursor="pointer"
         opacity="1"
         transform-origin="-18.5625px 0px"
         style="transform: translateY(340px); transform-origin: -18.5625px 0px;"
      >
        <circle class="grouptree__node  grouptree__node--withchildren"
                r="15"
        >
        </circle>
        <text class="grouptree__nodelabel"
              dominant-baseline="middle"
              x="-21"
              text-anchor="end"
        >
          TRE_1
        </text>
        <text class="grouptree__pressurelabel"
              text-anchor="middle"
              y="-.05em"
        >
          5
        </text>
        <text class="grouptree__pressureunit"
              y=".04em"
              dominant-baseline="text-before-edge"
              text-anchor="middle"
        >
          Bar
        </text>
        <title>
          Pressure: 5 Bar
        </title>
      </g>
      <g class="node"
         opacity="1"
         transform-origin="24.34228515625px 0px"
         style="transform: translateX(748px) translateY(170px); transform-origin: 24.3423px 0px;"
      >
        <circle class="grouptree__node"
                r="15"
        >
        </circle>
        <text class="grouptree__nodelabel"
              dominant-baseline="middle"
              x="21"
              text-anchor="start"
        >
          TRE_1_1
        </text>
        <text class="grouptree__pressurelabel"
              text-anchor="middle"
              y="-.05em"
        >
          20
        </text>
        <text class="grouptree__pressureunit"
              y=".04em"
              dominant-baseline="text-before-edge"
              text-anchor="middle"
        >
          Bar
        </text>
        <title>
          Pressure: 20 Bar
Bottom Hole Pressure: 11 N/m2
: 12 ?
        </title>
      </g>
      <g class="node"
         opacity="1"
         transform-origin="24.34228515625px 0px"
         style="transform: translateX(748px) translateY(510px); transform-origin: 24.3423px 0px;"
      >
        <circle class="grouptree__node"
                r="15"
        >
        </circle>
        <text class="grouptree__nodelabel"
              dominant-baseline="middle"
              x="21"
              text-anchor="start"
        >
          TRE_1_2
        </text>
        <text class="grouptree__pressurelabel"
              text-anchor="middle"
              y="-.05em"
        >
          22
        </text>
        <text class="grouptree__pressureunit"
              y=".04em"
              dominant-baseline="text-before-edge"
              text-anchor="middle"
        >
          Bar
        </text>
        <title>
          Pressure: 22 Bar
: 10 ?
        </title>
      </g>
    </g>
  </svg>
</div>
`;

exports[`GroupTreePlot/Demo Resizable smoke-test 2`] = `
<div style="height: 700px; width: 700px;">
  <button>
    Randomize size
  </button>
  <div style="margin-top: 0.25rem; background: rgba(0, 0, 255, 0.2); height: 400px; width: 400px; transition: height 200ms ease 0s, width ease 0s;">
    <svg height="100%"
         width="100%"
    >
      <g transform="translate(70,10)">
        <g>
          <path id=":r8:"
                class="link grouptree_link grouptree_link__waterrate"
                stroke-dasharray="none"
                stroke-opacity="1"
                stroke-width="57.99999999999999"
                d="
        M 0 190         C             105 190,             105 95,             210 95
        "
          >
            <title>
              Water Rate: 20 m3/s
Oil Rate: 30 m3/s
Gas Rate: 40 m3/s
Water Injection Rate: 50 m3/s
Gas Injection Rate: 60 m3/s
            </title>
          </path>
          <text dominant-baseline="central"
                text-anchor="middle"
                fill-opacity="1"
          >
            <textPath class="edge_info_text"
                      startoffset="50%"
                      href="#:r8:"
            >
              VFP11
            </textPath>
          </text>
        </g>
        <g>
          <path id=":ra:"
                class="link grouptree_link grouptree_link__waterrate"
                stroke-dasharray="none"
                stroke-opacity="1"
                stroke-width="72"
                d="
        M 0 190         C             105 190,             105 285,             210 285
        "
          >
            <title>
              Water Rate: 25 m3/s
Oil Rate: 35 m3/s
Gas Rate: 45 m3/s
Water Injection Rate: 55 m3/s
Gas Injection Rate: 65 m3/s
            </title>
          </path>
          <text dominant-baseline="central"
                text-anchor="middle"
                fill-opacity="1"
          >
            <textPath class="edge_info_text"
                      startoffset="50%"
                      href="#:ra:"
            >
              VFP12
            </textPath>
          </text>
        </g>
        <g class="node"
           cursor="pointer"
           opacity="1"
           transform-origin="-18.5625px 0px"
           style="transform: translateY(190px); transform-origin: -18.5625px 0px;"
        >
          <circle class="grouptree__node  grouptree__node--withchildren"
                  r="15"
          >
          </circle>
          <text class="grouptree__nodelabel"
                dominant-baseline="middle"
                x="-21"
                text-anchor="end"
          >
            TRE_1
          </text>
          <text class="grouptree__pressurelabel"
                text-anchor="middle"
                y="-.05em"
          >
            5
          </text>
          <text class="grouptree__pressureunit"
                y=".04em"
                dominant-baseline="text-before-edge"
                text-anchor="middle"
          >
            Bar
          </text>
          <title>
            Pressure: 5 Bar
          </title>
        </g>
        <g class="node"
           opacity="1"
           transform-origin="24.34228515625px 0px"
           style="transform: translateX(210px) translateY(95px); transform-origin: 24.3423px 0px;"
        >
          <circle class="grouptree__node"
                  r="15"
          >
          </circle>
          <text class="grouptree__nodelabel"
                dominant-baseline="middle"
                x="21"
                text-anchor="start"
          >
            TRE_1_1
          </text>
          <text class="grouptree__pressurelabel"
                text-anchor="middle"
                y="-.05em"
          >
            20
          </text>
          <text class="grouptree__pressureunit"
                y=".04em"
                dominant-baseline="text-before-edge"
                text-anchor="middle"
          >
            Bar
          </text>
          <title>
            Pressure: 20 Bar
Bottom Hole Pressure: 11 N/m2
: 12 ?
          </title>
        </g>
        <g class="node"
           opacity="1"
           transform-origin="24.34228515625px 0px"
           style="transform: translateX(210px) translateY(285px); transform-origin: 24.3423px 0px;"
        >
          <circle class="grouptree__node"
                  r="15"
          >
          </circle>
          <text class="grouptree__nodelabel"
                dominant-baseline="middle"
                x="21"
                text-anchor="start"
          >
            TRE_1_2
          </text>
          <text class="grouptree__pressurelabel"
                text-anchor="middle"
                y="-.05em"
          >
            22
          </text>
          <text class="grouptree__pressureunit"
                y=".04em"
                dominant-baseline="text-before-edge"
                text-anchor="middle"
          >
            Bar
          </text>
          <title>
            Pressure: 22 Bar
: 10 ?
          </title>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
